streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java

```diff
@@ -583,6 +627,172 @@ public void advanceWallClockTime(final long advanceMs) {
+            return null;
+        }
+        return outputRecords.poll();
+    }
+
+    /**
+     * Read the next record from the given topic.
+     * These records were output by the topology during the previous calls to {@link #pipeInput(ConsumerRecord)}.
+     *
+     * @param topic             the name of the topic
+     * @param keyDeserializer   the deserializer for the key type
+     * @param valueDeserializer the deserializer for the value type
+     * @param <K> the key type
+     * @param <V> the value type
+     * @return the next record on that topic, or {@code null} if there is no record available
+     */
+    @SuppressWarnings("WeakerAccess")
+    <K, V> TestRecord<K, V> readRecord(final String topic,
+                                         final Deserializer<K> keyDeserializer,
+                                         final Deserializer<V> valueDeserializer) {
+        final Queue<? extends ProducerRecord<byte[], byte[]>> outputRecords = getRecordsQueue(topic);
+        if (outputRecords == null) {
+            throw new NoSuchElementException("Uninitialized topic: " + topic);
+        }
+        final ProducerRecord<byte[], byte[]> record = outputRecords.poll();
+        if (record == null) {
+            throw new NoSuchElementException("Empty topic: " + topic);
+        }
+        final K key = keyDeserializer.deserialize(record.topic(), record.key());
+        final V value = valueDeserializer.deserialize(record.topic(), record.value());
+        return new TestRecord<>(key, value, record.headers(), record.timestamp());
+    }
+
+    /**
+     * Serialize an input recond and send on the specified topic to the topology and then
+     * commit the messages.
+     *
+     * @param topic           the name of the topic
+     * @param record          TestRecord to be send
+     * @param keySerializer   the key serializer
+     * @param valueSerializer the value serializer
+     * @param <K> the key type
+     * @param <V> the value type*
+     * @param time            timestamp to override the record timestamp
+     */
+    @SuppressWarnings("WeakerAccess")
+    <K, V> void pipeRecord(final String topic,
+                           final TestRecord<K, V> record,
+                           final Serializer<K> keySerializer,
+                           final Serializer<V> valueSerializer,
+                           final Instant time) {
+        final byte[] serializedKey = keySerializer.serialize(topic, record.headers(), record.key());
+        final byte[] serializedValue = valueSerializer.serialize(topic, record.headers(), record.value());
+        //final Long timestamp = (time != null) ? time.toEpochMilli() : record.timestamp();
+        long timestamp = 0;
+        if (time != null)
+            timestamp = time.toEpochMilli();
+        else if (record.timestamp() != null)
+            timestamp = record.timestamp();
+
+        pipeRecord(topic, timestamp, serializedKey, serializedValue, record.headers());
```

Comment:
if `time == null && record.timestamp() == null` we pass `timestamp==0`; is this intended? Sounds like an error case to me (should we throw an exception, or can this never happen anyway?)

Reply:
I needed to modify like that to get some old test to work. It might be some non valid test.

Reply:
Well. Overwriting the timestamp is fine, however, if we don't pass in `final Instant time`, we should just pass `record.timestamp()` into `pipeRecord()` -- I checked out the code applied the following change an run test (and they passed---hence, I think it fine to be more strict)

```
long timestamp;
if (time != null) {
    timestamp = time.toEpochMilli();
} else if (record.timestamp() != null) {
    timestamp = record.timestamp();
} else {
    throw new IllegalStateException("Provided `TestRecord` does not have a timestamp and no timestamp overwrite was provided via `time` parameter.");
}
```

Reply:
The some tests failing if making change like this.